src/embedded_flash/helper.rs

```diff
@@ -0,0 +1,282 @@
+pub struct ModRange {
+    start: usize,
+    length: usize,
+}
+
+impl ModRange {
+    /// Returns a new range of given start and length.
+    ///
+    /// The stored length can be shorter than what is passed in, so that the range does not include
+    /// values bigger than `usize::MAX`.
+    pub fn new(start: usize, mut length: usize) -> ModRange {
+        if length > 0 {
+            length = start.saturating_add(length - 1) - start + 1;
+        }
+        ModRange { start, length }
+    }
+
+    /// Create a new empty range.
+    pub fn new_empty() -> ModRange {
+        ModRange::new(0, 0)
+    }
+
+    /// Returns the start of the range.
+    pub fn start(&self) -> usize {
+        self.start
+    }
+
+    /// Returns the length of the range.
+    pub fn length(&self) -> usize {
+        self.length
+    }
+
+    /// Returns whether this range contains any addresses.
+    pub fn is_empty(&self) -> bool {
+        self.length == 0
+    }
+
+    /// Returns the disjoint union with the other range, if is consecutive.
+    ///
+    /// Appending empty ranges is not possible.
+    /// Appending to the empty range returns the other range.
+    pub fn append(&self, other: ModRange) -> Option<ModRange> {
+        if self.is_empty() {
+            return Some(other);
+        }
+        if other.is_empty() {
+            return None;
+        }
+        if self.start >= other.start {
+            return None;
+        }
+        if self.length != other.start - self.start {
+            return None;
+        }
+        let new_length = self.length.checked_add(other.length);
+        new_length.map(|l| ModRange::new(self.start, l))
+    }
+
+    /// Returns whether the given address is inside the range.
+    pub fn contains(&self, x: usize) -> bool {
+        // We want to check the 2 following inequalities:
+        // (1) `start <= x`
+        // (2) `x < start + length`
+        // However, the second one may overflow written as is. Using (1), we rewrite to:
+        // (3) `x - start <= length`
+        self.start <= x && x - self.start < self.length
+    }
+
+    /// Returns whether the given range is fully contained, i.e. their cut equals the parameter.
+    pub fn contains_range(&self, range: &ModRange) -> bool {
+        if range.is_empty() {
+            return true;
+        }
+        let end = range.start + (range.length - 1);
+        self.contains(range.start) && self.contains(end)
+    }
+
+    /// Returns an iterator for all contained numbers that are divisible by the modulus.
+    pub fn aligned_iter(&self, modulus: usize) -> impl Iterator<Item = usize> {
+        let remainder = self.start % modulus;
+        // Saturating first and limit in case they are usize::MAX.
+        let first = if remainder == 0 {
+            self.start
+        } else {
+            (self.start - remainder).saturating_add(modulus)
+        };
+        let limit = self.start.saturating_add(self.length);
```

Comment:
Okay, I should have documented that I ignore MAX, motivated by application, i.e. no page will start at MAX. It should be okay to add that in the documentation that MAX is not returned, instead of changing the return type?

Reply:
Actually looking at this, it looks quite complicated for what it does. I'm not sure it's worth dealing with MAX that much because we will never go close to it (embedded devices won't get 4G of flash before they get to 64 bits).

It looks like only doing the arithmetic in a way to avoid overflow should be enough:
- `new` would not truncate
- `append` would only saturate the length
- the iteration could look like: `(start..).take(length).step_by(modulo)`

What do you think?

Reply:
Truncation in `new` allows the other functions to call the standard operators, like 
```
let end = range.start + (range.length - 1);
```

in `contains_range`. Saturation and checking in `append` isn't much of a difference to me, just fails a bit earlier if you have a typo. I tried to build it robust enough to handle external input, expecting offsets to be passed through CTAP directly from the client.

Reply:
`contains_range` can be written more simply without any assumptions (and then `contains` removed because we don't need it):

```rust
self.start <= other.start && other.length <= self.length &&
    other.start - self.start <= self.length - other.length
```

Yes, I agree we should never trust CTAP input. But I don't think too many checks/invariants are helping, because they complexify the code all the time, for a debugging that might never happen and wouldn't be that much more complicated without (printf debugging). I think most important is that the code is kept simple (thus readable and probably also extensible, because there is less code that is not obvious whether it is actually needed or just here for debugging). Debugging asserts can be added because they are easy to filter out. If extra code for robustness against the language is needed (to prevent against bugs in the compiler or the underlying hardware), then yes we have to add that but only if it's worth it.

Reply:
I added an `is_empty()` check because empty ranges should always be contained in anything. Otherwise, I agree to the code complexity! `contains` is removed, and therefore `new` can be simplified.